File: C++ Codes/vertex_cover_spoj_bfs.cpp
#include<bits/stdc++.h>

using namespace std;

const int N = 1e5 + 1;

vector<int> gr[N];
int dp[N][2];
int vis[N];

int main()
{
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);

	int n, root;
	cin >> n;

	// corrected n-1 as edge count
	for (int i = 0; i < n - 1; i++) {
		int x, y;
		cin >> x >> y;
		gr[x].push_back(y);
		gr[y].push_back(x);
	}

	queue<int> Q;

	for (int i = 1; i <= n; i++) {
		if (gr[i].size() == 1) Q.push(i);
	}

	while (!Q.empty()) {

		int cur = Q.front();
		Q.pop();

		vis[cur] = 1;
		root = cur;

		dp[cur][0] = 0;
		dp[cur][1] = 1;

		for (auto x : gr[cur]) {
			if (vis[x]) {
				// child nodes
				dp[cur][0] += dp[x][1];
				dp[cur][1] += min(dp[x][0], dp[x][1]);
			}
			else {
				// parent node
				Q.push(x);
			}
		}

	}
	// the node which is visited in the end
	cout << min(dp[root][0], dp[root][1]);

	return 0;
}

File: Java Codes/VertexCoverSpojBFS.java
package DP;

import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class VertexCoverSpojBFS {


    static final int N = (int) (1e5 + 1);

    static ArrayList<Integer>[] gr = new ArrayList[N];
    static int[][] dp = new int[N][2];

    static int[] vis = new int[N];

    public static void main(String[] args) {

        Scanner scn = new Scanner(System.in);
        int n = scn.nextInt(), root = -1;

        for (int i = 0; i < n - 1; i++) {
            int x = scn.nextInt(), y = scn.nextInt();

            gr[x].add(y);
            gr[y].add(x);
        }

        Queue<Integer> Q = new LinkedList<>();

        for (int i = 1; i <= n; i++) {
            if (gr[i].size() == 1) Q.add(i);
        }

        while (!Q.isEmpty()) {

            int cur = Q.poll();

            vis[cur] = 1;
            root = cur;

            dp[cur][0] = 0;
            dp[cur][1] = 1;

            for (int x : gr[cur]) {
                if (vis[x] == 1) {
                    // child nodes
                    dp[cur][0] += dp[x][1];
                    dp[cur][1] += Math.min(dp[x][0], dp[x][1]);
                }
                else {
                    // parent node
                    Q.add(x);
                }
            }

        }
        // the node which is visited in the end
        System.out.println(Math.min(dp[root][0], dp[root][1]));
    }
}
